src/lib.rs

```diff
@@ -11,55 +12,87 @@ extern crate schnorrkel;
 // forked at commit eff430ddc3090f56317c80654208b8298ef7ab3f
 
 use schnorrkel::{
-	derive::{ChainCode, Derivation, CHAIN_CODE_LENGTH},
-	Keypair, MiniSecretKey, PublicKey, SecretKey, Signature,
+    derive::{ChainCode, Derivation, CHAIN_CODE_LENGTH},
+    Keypair, MiniSecretKey, PublicKey, SecretKey, Signature,
+    context::signing_context,
+    vrf::{VRFOutput, VRFProof},
 };
 
 use std::ptr;
 use std::slice;
+use std::fmt::Write;
+
+// Absent on OS X
+pub const PT_NULL: u32 = 0;
 
 // We must make sure that this is the same as declared in the substrate source code.
 const SIGNING_CTX: &'static [u8] = b"substrate";
 
 /// ChainCode construction helper
 fn create_cc(data: &[u8]) -> ChainCode {
-	let mut cc = [0u8; CHAIN_CODE_LENGTH];
+    let mut cc = [0u8; CHAIN_CODE_LENGTH];
 
-	cc.copy_from_slice(&data);
+    cc.copy_from_slice(&data);
 
-	ChainCode(cc)
+    ChainCode(cc)
 }
 
 /// Keypair helper function.
 fn create_from_seed(seed: &[u8]) -> Keypair {
-	match MiniSecretKey::from_bytes(seed) {
-		Ok(mini) => return mini.expand_to_keypair(),
-		Err(_) => panic!("Provided seed is invalid."),
-	}
+    match MiniSecretKey::from_bytes(seed) {
+        Ok(mini) => return mini.expand_to_keypair(),
+        Err(_) => panic!("Provided seed is invalid."),
+    }
 }
 
 /// Keypair helper function.
 fn create_from_pair(pair: &[u8]) -> Keypair {
-	match Keypair::from_bytes(pair) {
-		Ok(pair) => return pair,
-		Err(_) => panic!(format!("Provided pair is invalid: {:?}", pair)),
-	}
+    match Keypair::from_bytes(pair) {
+        Ok(pair) => return pair,
+        Err(_) => panic!(format!("Provided pair is invalid: {:?}", pair)),
+    }
 }
 
 /// PublicKey helper
 fn create_public(public: &[u8]) -> PublicKey {
-	match PublicKey::from_bytes(public) {
-		Ok(public) => return public,
-		Err(_) => panic!("Provided public key is invalid."),
-	}
+    match PublicKey::from_bytes(public) {
+        Ok(public) => return public,
+        Err(_) => panic!("Provided public key is invalid."),
+    }
 }
 
 /// SecretKey helper
 fn create_secret(secret: &[u8]) -> SecretKey {
-	match SecretKey::from_bytes(secret) {
-		Ok(secret) => return secret,
-		Err(_) => panic!("Provided private key is invalid."),
-	}
+    match SecretKey::from_bytes(secret) {
+        Ok(secret) => return secret,
+        Err(_) => panic!("Provided private key is invalid."),
+    }
+}
+
+/**
+ * Allocate a string in memory with prefix \param prefix and suffix, which is the error message of \param e
+ */
+unsafe fn allocate_error_string_from(prefix: &str, e: &schnorrkel::SignatureError) -> *const libc::c_char {
+    let mut err_msg = String::new();
+    write!(&mut err_msg, "{}: {}", prefix, e);
+    let ptr = libc::malloc(err_msg.len());
+    let out_str = slice::from_raw_parts_mut(ptr as *mut libc::c_char, err_msg.len());
+    for i in 0..(err_msg.len()) {
+        out_str[i] = err_msg.as_bytes()[i] as i8;
+    }
+    out_str.as_ptr() as *const _
```

Comment:
people will always forget to free error messages, so it is better to NOT do memory allocation at all. 

it is better to just have enum and short comments alongside of each error (remember that 0 is a success)